massa-protocol-worker/src/protocol_network.rs

```diff
@@ -0,0 +1,426 @@
+            let block_info = match info_wanted {
+                AskForBlocksInfo::Info => ReplyForBlocksInfo::Info(operations_ids),
+                AskForBlocksInfo::Operations(op_ids) => {
+                    // Mark the node as having the block.
+                    node_info.insert_known_blocks(
+                        &[*hash],
+                        true,
+                        Instant::now(),
+                        self.config.max_node_known_blocks_size,
+                    );
+
+                    // Send only the missing operations.
+                    let needed_ops = operations_ids
+                        .into_iter()
+                        .filter(|id| op_ids.contains(id))
+                        .collect();
+                    ReplyForBlocksInfo::Operations(needed_ops)
+                }
+            };
+            all_blocks_info.push((*hash, block_info));
+        }
+        self.network_command_sender
+            .send_block_info(from_node_id, all_blocks_info)
+            .await
+            .map_err(|_| {
+                ProtocolError::ChannelError("send block info network command send failed".into())
+            })
+    }
+
+    /// On block information received, manage when we get a list of operations.
+    /// Ask for the missing operations that are not in the `checked_operations` cache variable.
+    ///
+    /// # Ban
+    /// Start compute the operations serialized total size with the operation we know.
+    /// Ban the node if the operations contained in the block overflow the max size. We don't
+    /// forward the block to the graph in that case.
+    ///
+    /// # Parameters:
+    /// - `from_node_id`: Node which sent us the information.
+    /// - `BlockId`: ID of the related operations we received.
+    /// - `operation_ids`: IDs of the operations contained by the block.
+    ///
+    /// # Result
+    /// return an error if stopping asking block failed. The error should be forwarded at the
+    /// root. todo: check if if make panic.
+    async fn on_block_operation_list_received(
+        &mut self,
+        from_node_id: NodeId,
+        block_id: BlockId,
+        operation_ids: OperationIds,
+    ) -> Result<(), ProtocolError> {
+        match self.block_wishlist.get(&block_id) {
+            Some(AskForBlocksInfo::Info) => {}
+            _ => return Ok(()),
+        }
+        let mut info = match self.checked_headers.get_mut(&block_id) {
+            Some(info) => info,
+            _ => {
+                warn!("Missing block header for {}", block_id);
+                return Ok(());
+            }
+        };
+        let mut total_hash: Vec<u8> = vec![];
+        operation_ids.iter().for_each(|op_id| {
+            let op_hash = op_id.hash().into_bytes();
+            total_hash.extend(op_hash);
+        });
+
+        // Check operation_list against expected operations hash from header.
+        if info.header.content.operation_merkle_root == Hash::compute_from(&total_hash) {
```

Comment:
let's rename operation_merkle_root to operation_global_hash

Reply:
to avoid 1 level of indentation, use:

```
if info.header.content.operation_merkle_root != Hash::compute_from(&total_hash) {
    let _ = self.ban_node(&from_node_id).await;
    return Ok(());
}
```

Reply:
@gterzian